windows/CodePush.cs

```diff
@@ -0,0 +1,668 @@
+            if (pendingUpdateString == null)
+            {
+                return null;
+            }
+
+            try
+            {
+                return JObject.Parse(pendingUpdateString);
+            }
+            catch (Exception)
+            {
+                // Should not happen.
+                CodePushUtils.log("Unable to parse pending update metadata " + pendingUpdateString +
+                        " stored in SharedPreferences");
+                return null;
+            }
+        }
+
+        private void InitializeUpdateAfterRestart()
+        {
+            JObject pendingUpdate = GetPendingUpdate();
+            if (pendingUpdate != null)
+            {
+                didUpdate = true;
+                bool updateIsLoading = (bool)pendingUpdate[PENDING_UPDATE_IS_LOADING_KEY];
+                if (updateIsLoading)
+                {
+                    // Pending update was initialized, but notifyApplicationReady was not called.
+                    // Therefore, deduce that it is a broken update and rollback.
+                    CodePushUtils.log("Update did not finish loading the last time, rolling back to a previous version.");
+                    needToReportRollback = true;
+                    RollbackPackage().Wait();
+                }
+                else
+                {
+                    // Clear the React dev bundle cache so that new updates can be loaded.
+                    if (mainPage.UseDeveloperSupport)
+                    {
+                        ClearReactDevBundleCache().Wait();
+                    }
+                    // Mark that we tried to initialize the new update, so that if it crashes,
+                    // we will know that we need to rollback when the app next starts.
+                    SavePendingUpdate((string)pendingUpdate[PENDING_UPDATE_HASH_KEY], /* isLoading */true);
+                }
+            }
+        }
+
+        private bool IsFailedHash(string packageHash)
+        {
+            JArray failedUpdates = GetFailedUpdates();
+            if (packageHash != null)
+            {
+                foreach (JObject failedPackage in failedUpdates)
+                {
+                    string failedPackageHash = (string)failedPackage[PACKAGE_HASH_KEY];
+                    if (packageHash.Equals(failedPackageHash))
+                    {
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        private bool IsPendingUpdate(string packageHash)
+        {
+            JObject pendingUpdate = GetPendingUpdate();
+            return pendingUpdate != null &&
+                    !(bool)pendingUpdate[PENDING_UPDATE_IS_LOADING_KEY] &&
+                    (packageHash == null || ((string)pendingUpdate[PENDING_UPDATE_HASH_KEY]).Equals(packageHash));
+        }
+
+        private void RemoveFailedUpdates()
```

Comment:
this and RemovePendingUpdate are nearly identical but with a different key...not sure how helpful it would be to make a helper function but if there were 3 functions here I would definitely do it